src/node_union_bytes.h

```diff
@@ -4,50 +4,74 @@
+class StaticExternalByteResource : public Base {
+  static_assert(sizeof(IChar) == sizeof(Char),
+                "incompatible interface and internal pointers");
+
+ public:
+  explicit StaticExternalByteResource(const Char* data,
+                                      size_t length,
+                                      std::shared_ptr<void> owning_ptr)
+      : data_(data), length_(length), owning_ptr_(owning_ptr) {}
+
+  const IChar* data() const override {
+    return reinterpret_cast<const IChar*>(data_);
+  }
+  size_t length() const override { return length_; }
+
+  void Dispose() override {
+    // We ignore Dispose calls from V8, even if we "own" a resource via
+    // owning_ptr_. All instantiations of this class are static or owned by a
+    // static map, and will be destructed when static variables are destructed.
+  }
+
+  StaticExternalByteResource(const StaticExternalByteResource&) = delete;
+  StaticExternalByteResource& operator=(const StaticExternalByteResource&) =
+      delete;
+
+ private:
+  const Char* data_;
+  const size_t length_;
+  std::shared_ptr<void> owning_ptr_;
+};
+
+using StaticExternalOneByteResource =
+    StaticExternalByteResource<uint8_t,
+                               char,
+                               v8::String::ExternalOneByteStringResource>;
+using StaticExternalTwoByteResource =
+    StaticExternalByteResource<uint16_t,
+                               uint16_t,
+                               v8::String::ExternalStringResource>;
+
 // Similar to a v8::String, but it's independent from Isolates
 // and can be materialized in Isolates as external Strings
 // via ToStringChecked.
 class UnionBytes {
  public:
-  UnionBytes(const uint16_t* data, size_t length)
-      : one_bytes_(nullptr), two_bytes_(data), length_(length) {}
-  UnionBytes(const uint8_t* data, size_t length)
-      : one_bytes_(data), two_bytes_(nullptr), length_(length) {}
-  template <typename T>  // T = uint8_t or uint16_t
-  explicit UnionBytes(std::shared_ptr<std::vector</*const*/ T>> data)
-      : UnionBytes(data->data(), data->size()) {
-    owning_ptr_ = data;
-  }
+  explicit UnionBytes(StaticExternalOneByteResource* one_byte_resource)
+      : one_byte_resource_(one_byte_resource), two_byte_resource_(nullptr) {}
+  explicit UnionBytes(StaticExternalTwoByteResource* two_byte_resource)
+      : one_byte_resource_(nullptr), two_byte_resource_(two_byte_resource) {}
 
   UnionBytes(const UnionBytes&) = default;
   UnionBytes& operator=(const UnionBytes&) = default;
   UnionBytes(UnionBytes&&) = default;
   UnionBytes& operator=(UnionBytes&&) = default;
 
-  bool is_one_byte() const { return one_bytes_ != nullptr; }
-  const uint16_t* two_bytes_data() const {
-    CHECK_NOT_NULL(two_bytes_);
-    return two_bytes_;
-  }
-  const uint8_t* one_bytes_data() const {
-    CHECK_NOT_NULL(one_bytes_);
-    return one_bytes_;
-  }
+  bool IsOneByte() const { return one_byte_resource_ != nullptr; }
```

Comment:
For cheap operations we use the snake case for the method names.

```suggestion
  bool is_one_byte() const { return one_byte_resource_ != nullptr; }
```